src/vanilla/store.ts

```diff
@@ -200,8 +194,11 @@ const registerBatchAtom = (
 ) => {
   if (!batch.D.has(atom)) {
     batch.D.set(atom, new Set())
-    addBatchFuncMedium(batch, () => {
-      atomState.m?.l.forEach((listener) => addBatchFuncMedium(batch, listener))
+    addBatchFunc(batch, 'H', () => {
+      atomState.m?.l.forEach(
+        (listener: (() => void) & { INTERNAL_priority?: BatchPriority }) =>
+          addBatchFunc(batch, listener.INTERNAL_priority || 'M', listener),
+      )
     })
```

Comment:
1. Raise the priority of the listener insertion function so that it can insert to the high priority channel. Synchronous effects should run after recomputeDependents and before all listeners and mount hooks.
2. Determine the listener priority by reading an internal property tag on the listener.